gateway/app/drivers/qodProvisioning/nef.py

```diff
@@ -0,0 +1,290 @@
+from app.schemas.afSessionWithQos import (
+    AsSessionWithQoSSubscription,
+    UserPlaneNotificationData,
+)
+from app.schemas.commonData import (
+    FlowInfo,
+    Link,
+    UserPlaneEvent,
+)
+from app.schemas.qodProvisioning import (
+    CloudEvent,
+    CreateProvisioning,
+    Datacontenttype,
+    ProvisioningInfo,
+    RetrieveProvisioningByDevice,
+    Specversion,
+    Status,
+    StatusInfo,
+    Type,
+)
+from app.settings import settings
+import hashlib
+
+_prefix_info = "qodprovisioninginfo"
+_prefix_gateway_nef = "qodprovisioningnef"
+_prefix_device = "qodDevice"
+LOG = logging.getLogger(__name__)
+
+
+class NEFQoDProvisioningInterface(QoDProvisioningInterface):
+    def __init__(self) -> None:
+        super().__init__()
+        self.httpx_client = httpx.AsyncClient()
+        self.redis = get_redis()
+
+    async def create_provisioning(self, req: CreateProvisioning) -> ProvisioningInfo:
+        provisioning_id = uuid.uuid4()
+
+        payload = AsSessionWithQoSSubscription(
+            supportedFeatures="",
+            notificationDestination=Link(
+                f"{settings.qod_provisioning.gateway_url_callback}/{provisioning_id}"
+            ),
+            flowInfo=[
+                FlowInfo(
+                    flowId=0,
+                    flowDescriptions=[
+                        "permit in ip any to any",
+                        "permit out ip any to any",
+                    ],
+                )
+            ],
+            qosReference=req.qosProfile,
+        )
+
+        if req.device:
+            payload.ueIpv4Addr = req.device.ipv4Address.publicAddress
+            payload.ueIpv6Addr = req.device.ipv6Address
+            payload.gpsi = (
+                f"msisdn-{req.device.phoneNumber}" if req.device.phoneNumber else None
+            )
+
+        res = await self.httpx_client.post(
+            f"{settings.qod_provisioning.nef_url}/{settings.qod_provisioning.af_id}/subcriptions",
+            content=payload.model_dump_json(),
+        )
+
+        if res.status_code == 404:
+            raise ResourceNotFound()
+
+        subcription_result = AsSessionWithQoSSubscription.model_validate_json(
+            res.content
+        )
+
+        nef_sub_id = self._get_subscription_id_from_subscription_url(
+            subcription_result.self
+        )
+
+        response = ProvisioningInfo(
+            device=req.device,
+            qosProfile=req.qosProfile,
+            status=Status.REQUESTED,
+            sink=req.sink,
+            provisioningId=provisioning_id,
+            startedAt=datetime.datetime.now(),
+        )
+
+        key = f"{_prefix_info}:{id}"
+        self.redis.set(key, response.model_dump_json())
+
+        key = f"{_prefix_gateway_nef}:{id}"
+        self.redis.set(key, nef_sub_id)
+
+        if req.device:
+            device_hash = hashlib.sha256(
+                req.device.model_dump_json().encode("UTF-8")
+            ).hexdigest()
```

Comment:
This won't work, you need a key for each different identifier, because the client might supply different identifiers for the create request and the get by device request.